remove use of state in nodejs app

tysonnorris opened the pull request:
To allow possibility of concurrent activation processing at the invoker, the "state" handling in the nodejs action container needs to be removed; Based on some discussion in the past, this was leftover from a legacy impl that is no longer susceptible to concurrent activation processing since fixing of bugs and move to actor based activation processing in the invoker. 

I don't have a reference to that discussion, but please add comments here if you have concerns.

rabbah commented:
Indeed this is legacy defensive code but it has been observed to still catch some lifecycle management bugs in recent memory. It is also a unique feature of this runtime and not other runtime proxies.

tysonnorris commented:
@csantanapr @rabbah @markusthoemmes any objections to merging this as is? (after rebase)

csantanapr commented:
@tysonnorris can you add a unit test that runs a few concurrent actions and that all of them eventually finish. You can run actions that have different sleeps and that they finished in reverse order or invoke. 
When I say unit tests I mean, just using the ActionContsiner style of tests with no whisk/controller system.

tysonnorris commented:
Working on tests - they won't run till https://github.com/apache/incubator-openwhisk/pull/3687 (which could possibly supplant https://github.com/apache/incubator-openwhisk/pull/2795, if useful) is merged, since the tests require changes to `ActionContainer` and friends to support this type of testing. 

Already added a concurrency test, will add some varied parameter ones.

rabbah commented:
@tysonnorris could we factor out the changes to `ActionContainer` to break the cyclic dependence? I looked over those specific change and think we can merge that much faster which then permits these tests to run.

rabbah commented on tests/src/test/scala/runtime/actionContainers/NodeJsActionContainerTests.scala:
```diff
@@ -547,4 +547,32 @@ abstract class NodeJsActionContainerTests extends BasicActionRunnerTests with Ws
+          |     console.log("interleave me");
+          |     return new Promise(function(resolve, reject) {
+          |       setTimeout(function() {
+          |         resolve({ done: true });
```
i suggest changing the return to be the identity (return `args`) and passing in two different values to make sure the response is say either `{a: "A"}` or `{b: "B"}` and that both are received.

tysonnorris commented:
@rabbah you mean to separate the ActionContainer changes into its own PR?

rabbah commented:
Yes.

tysonnorris commented:
@csantanapr let me know if these changes address your concerns.

csantanapr commented on tests/src/test/scala/runtime/actionContainers/NodeJsActionContainerTests.scala:
```diff
@@ -547,4 +547,42 @@ abstract class NodeJsActionContainerTests extends BasicActionRunnerTests with Ws
+        runPayload(_)
+      })
+      payloads.foreach { a =>
+        //responses should contain(200, Some(JsObject("args" -> a)))
```
Why is this commented?

tysonnorris replied:
left from some previous debugging; fixed it

csantanapr commented on tests/src/test/scala/runtime/actionContainers/NodeJsActionContainerTests.scala:
```diff
@@ -547,4 +547,42 @@ abstract class NodeJsActionContainerTests extends BasicActionRunnerTests with Ws
+
+  it should "allow running activations concurrently" in {
+    val (out, err) = withNodeJsContainer { c =>
+      //this action will create a log entry, and only complete once all 3 activations have arrived and emitted logs
```
Why you mentioned 3 activations? I only see 2 below

csantanapr commented on tests/src/test/scala/runtime/actionContainers/NodeJsActionContainerTests.scala:
```diff
@@ -547,4 +547,42 @@ abstract class NodeJsActionContainerTests extends BasicActionRunnerTests with Ws
+
+    checkStreams(out, err, {
+      case (o, e) =>
+        o.replaceAll("\n", "") shouldBe "interleave meinterleave me"
```
I don't see how are you testing that actions are actually running at the same time, and also not in different containers.

tysonnorris commented:
@csantanapr what's the word on this?

csantanapr commented:
@tysonnorris this comment from @rabbah was ignored.

>@tysonnorris instead of removing entirely should it be feature toggled (it might be difficult to enable without a change in the container launch)?

The current behavior should be maintain, we saw a lot of field problems in production, with invokers and runtimes.

What do you think of adding an environment variable if present then allow the container to do concurrent invocations, if not present or `false` then leave it like it works today?

tysonnorris commented on tests/src/test/scala/runtime/actionContainers/NodeJsActionContainerTests.scala:
```diff
@@ -547,4 +547,48 @@ abstract class NodeJsActionContainerTests extends BasicActionRunnerTests with Ws
+           |     console.log("interleave me");
+           |     return new Promise(function(resolve, reject) {
+           |         setTimeout(function() {
+           |             if (global.count == 2) {
```
Me too - there is however something fishy about travis builds that failed if this was greater than 2; the same tests work great locally. Do you have any ideas on replicating that env? Should vagrant local be the closest thing? I assume this is some difference in the behavior of http client (in HttpUtils) that is somehow not allowing the extra connections (or maybe limited number of threads for concurrent future processing?) that only shows in travis...

tysonnorris commented:
I can add an env var to enable it.

tysonnorris commented:
@rabbah @csantanapr I have added an env var to enable concurrent activations, and updated tests to exec `akka.http.host-connection-pool.max-connections` activations concurrently. WDYT?
once https://github.com/apache/incubator-openwhisk/pull/3812 is merged, I would be ready to merge this PR as well.

rabbah commented on core/nodejsActionBase/src/service.js:
```diff
@@ -112,10 +113,14 @@ function NodeActionService(config) {
         if (status === Status.ready) {
-            setStatus(Status.running);
+            if (!ignoreRunStatus) {
+                setStatus(Status.running);
```
small nit --- how about just moving the predicate into `setStatus` itself 

```js
function setStatus(newStatus) {
  if (status !== Status.stopped && !ignoreRunStatus) {
    status = newStatus;
  }
}
```

tysonnorris replied:
I don't think that we want to ignore status in all cases, only when we are setting/testing `Status.running`. This seemed less confusing, with that in mind - WDYT?

rabbah commented on core/nodejsActionBase/app.js:
```diff
@@ -17,7 +17,8 @@
         'port': 8080,
-        'apiHost': process.env.__OW_API_HOST
+        'apiHost': process.env.__OW_API_HOST,
+        'allowConcurrent': process.env.OW_ALLOW_CONCURRENT
```
do you think we'll want to make `__OW_ALLOW_CONCURRENT` part of the activation context at some point? if so, using `__` prefix would be uniform.

tysonnorris replied:
I agree, but until that is done in OW core code, this seemed less confusing. I think there is some longer discussion about how that is done in OW core (runtime manifest?). I can go ahead and change it here in anticipation though.
